bsms.mediawiki

```diff
@@ -0,0 +1,169 @@
+# Whether the Signer's storage is tamper-proof.
+# Whether the Signer subsequently uses the multisig configuration to generate and verify receive and change addresses.
+
+An attacker who can modify the multisig configuration can steal or hold funds to ransom by duping the user into sending funds to the wrong address. An attacker who cannot modify the configuration but can learn about the keys and/or the configuration can monitor transactions in the wallet, resulting in loss of privacy.
+
+This proposal seeks to address concerns #1, #2 and #3: to mitigate the risk of tampering during the initial setup phase, and to define an interoperable multisig configuration format.
+
+Concerns #4 and #5 should be handled by Signers and are out of scope of this proposal.
+
+==Specification==
+
+===Prerequisites===
+This proposal assumes the parties in the multisig support [https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki BIP-0032], [https://github.com/bitcoin/bips/blob/master/bip-0322.mediawiki BIP-0322], [https://github.com/bitcoin/bitcoin/blob/master/doc/descriptors.md the descriptor language] and encryption.
+
+===Roles===
+====Coordinator====
+
+The Coordinator initiates the multisig setup. The Coordinator determines what type of multisig is used and the exact policy script. If encryption is enabled, the Coordinator also distributes a shared secret or shared secrets to the parties involved for secure communication. The Coordinator gathers information from the Signers to generate a descriptor record. The Coordinator distributes the descriptor record back to the Signers.
+
+====Signer====
+
+The Signer is a participating member in the multisig. Its responsibilities include providing its key record -- which contains an Extended Public Key (XPUB) -- to the Coordinator, verifying that its XPUB is included in the descriptor record and persisting the descriptor record in its storage.
+
+===Setup Process===
+
+====Round 1====
+
+=====Coordinator=====
+
+* The Coordinator creates a multisig wallet creation session. The Coordinator constructs the multisig script and its policy parameters, such as the total number of signers and the required number of signatures (<tt>M</tt> and <tt>N</tt>).
+* The session should expire after some time period determined by the Coordinator, e.g., 24 hours. The timeout allows the encryption key to have lower entropy.
+* If encryption is enabled, the Coordinator distributes a secret <tt>TOKEN</tt> to each Signer over a secure channel. The Signer can use the <tt>TOKEN</tt> to derive an <tt>ENCRYPTION_KEY</tt>. Refer to the Encryption section below for details on the <tt>TOKEN</tt>, the key derivation function and the encryption scheme. Depending on the use case, the Coordinator can decide whether to share one common <tt>TOKEN</tt> for all Signers, or to have one per Signer.
+* If encryption is disabled, <tt>TOKEN</tt> is set to <tt>0</tt>, and all the encryption/decryption steps below can be skipped.
```

Comment:
I think there's an issue with disabling the encryption.

if TOKEN doubles as an identifier of the multisig setup session, then setting it to 0 when encryption is disabled means that there will be no session identifier. The SIG of the xpub in the Signer->Coordinator record will commit to this '0' token, and then such `<0-token> <xpub> <SIG>` record can be reused in any other unencrypted multisig setup without access to corresponding privkey.

Reply:
This might matter because this way you can involve people into contracts without their consent (multisig script is a simple contract for controlling funds)

Reply:
> if TOKEN doubles as an identifier of the multisig setup session, then setting it to 0 when encryption is disabled means that there will be no session identifier.

In a way it is still a session, but a hardcoded session value of 0.

> The SIG of the xpub in the Signer->Coordinator record will commit to this '0' token, and then such <0-token> <xpub> <SIG> record can be reused in any other unencrypted multisig setup session without access to corresponding privkey.

Is there a way to prevent this though? I mean once you share an unencrypted XPUB, you are sort of opening yourself up to this scenario. And it is no worse than what we already have today. (The SIG doesn't really change the situation).

Reply:
The difference is that if SIG commits to unique session, then a malefactor won't be able to submit the xpub of the victim to the multisig setup without help from the coordinator of the said session.

Of course if the coordinator is the malefactor, there's no way to prevent this.

There could also be a mistake when an old xpub record is submitted to the coordinator. Since the record is not linked to the session, coordinator might include it in the final setup. The old xpub record might correspond to compromized or unavailable key. There's no third round of Signer->Coordinator where signer could confirm to Coordinator that the address can be started to be used, so the the funds could be sent to the address before the mistake is revealed.

I don't know how serious is these risks, but at least people willing to skip encryption should know about them.

The way to prevent this might be not using '0' as the 'no encryption' flag, and somehow indicate the 'no encryption' by other ways, and the TOKEN to always be a session token.

Reply:
> The difference is that if SIG commits to unique session, then a malefactor won't be able to submit the xpub of the victim to the multisig setup without help from the coordinator of the said session.

So one of the goals of this proposal is try to minimizing the requirements on the HWW side, and generally keeping things backward compatible. If there needs to be a TOKEN even when encryption is disabled, that means the user will always have to enter the TOKEN on the device, which increases friction.

> There could also be a mistake when an old xpub record is submitted to the coordinator. Since the record is not linked to the session, coordinator might include it in the final setup. The old xpub record might correspond to compromized or unavailable key. There's no third round of Signer->Coordinator where signer could confirm to Coordinator that the address can be started to be used, so the the funds could be sent to the address before the mistake is revealed.

I don't see how this is different from how one might accidentally reuse the same XPUB in two different multisig wallets currently though. The SIG needs to be committed to _something_. The alternative here is to not include the SIG if encryption is disabled - so there's no commitment, but I'm not sure if that's desirable, because the SIG does help even without encryption.

> The way to prevent this might be not using '0' as the 'no encryption' flag, and somehow indicate the 'no encryption' by other ways, and the TOKEN to always be a session token.

As mentioned above, that means that the user will have to set the TOKEN on the device even when encryption is disabled.

Reply:
> As mentioned above, that means that the user will have to set the TOKEN on the device even when encryption is disabled.

I see, and agree that this is good enough reason to have the 'zero-TOKEN' mode.